core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ResizeAllocationDecider.java

```diff
@@ -0,0 +1,100 @@
+
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.routing.RecoverySource;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.UnassignedInfo;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexNotFoundException;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.util.Set;
+
+/**
+ * An allocation decider that ensures we allocate the shards of a target index for resize operations next to the source primaries
+ * // TODO add tests!!
+ */
+public class ResizeAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "resize";
+
+    /**
+     * Initializes a new {@link ResizeAllocationDecider}
+     *
+     * @param settings {@link Settings} used by this {@link AllocationDecider}
+     */
+    public ResizeAllocationDecider(Settings settings) {
+        super(settings);
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocation) {
+        return canAllocate(shardRouting, null, allocation);
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        final UnassignedInfo unassignedInfo = shardRouting.unassignedInfo();
+        if (unassignedInfo != null && shardRouting.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS) {
+            // we only make decisions here if we have no unassigned info and we have to recover from another index ie. split / shrink
+            final IndexMetaData indexMetaData = allocation.metaData().getIndexSafe(shardRouting.index());
+            Index resizeSourceIndex = indexMetaData.getResizeSourceIndex();
+            assert resizeSourceIndex != null;
+            try {
+                IndexMetaData sourceIndexMetaData = allocation.metaData().getIndexSafe(resizeSourceIndex);
+                if (indexMetaData.getNumberOfShards() < sourceIndexMetaData.getNumberOfShards()) {
+                    // this only handles splits so far.
+                    return Decision.ALWAYS;
+                }
+                ShardId shardId = IndexMetaData.selectSplitShard(shardRouting.id(), sourceIndexMetaData, indexMetaData.getNumberOfShards());
+                ShardRouting sourceShardRouting = allocation.routingTable().shardRoutingTable(shardId).primaryShard();
+                if (sourceShardRouting.active() == false) {
+                    return allocation.decision(Decision.NO, NAME, "source primary shard [%s] is not active", sourceShardRouting.shardId());
+                }
+                if (node != null) { // we might get called from the 2 param canAllocate method..
```

Comment:
maybe we should also check if the ES version on the node is capable of splitting a shard.

Reply:
good point I will add that